### PR TITLE
tests: always make sure we return an LLB state

### PR DESCRIPTION
**Pull request by cpuguy83**
Before this we are returning errors to buildkit when there are test errors.
Or for that matter, if there is an error during evaluation of the stuff running up to the tests.
This causes issues with the debug workflow ("docker buildx dap") since there is no LLB state to attach to, so it can't set breakpoints properly.

With this change even when we get a solve error when evaluating state, we return the bad state so that the debugger can attach to it. When a test fails, we add a new LLB exec op which prints the joined test errors and exits non-zero so that the build will fail.

There's also some minor changes to make it simpler to discard any changes to the filesystem from running tests, or in the case of non-container outputs return the original state, while keeping the dependency chain in tact.

This is a simpler alternative to #898.
#898 is hitting some race condition in buildkit (or at least that's what it looks like).
Additionally this can actually be backported to older release branches as the changes are minor.

### CODE REVIEW

File: frontend/test_runner.go [hunk 1]
```diff
@@ -23,180 +25,218 @@ import (
...
-		}
-		if skip {
-			Warn(ctx, client, llb.Scratch(), "Tests skipped due to build-arg DALEC_SKIP_TESTS="+skipVar)
-			return nil
+// RunTests runs the tests defined in the spec against the given the input state.
+// The result of this is either the provided `finalState` or a state that errors when executed with the errors produced by the tests.
+// The input state should be a runnable container with all dependencies already installed.
+func RunTests(ctx context.Context, client gwclient.Client, spec *dalec.Spec, finalState llb.State, target string, platform *ocispecs.Platform) llb.StateOption {
+	return func(in llb.State) llb.State {
+		if skipVar := client.BuildOpts().Opts["build-arg:"+"DALEC_SKIP_TESTS"]; skipVar != "" {
```
Comment (MorrisLaw) [inline frontend/test_runner.go]: nit: we could just combing the string literals together

File: frontend/test_runner.go [hunk 2]
```diff
@@ -23,180 +25,218 @@ import (
...
+			skip, err := strconv.ParseBool(skipVar)
+			if err != nil {
+				err = errors.Wrapf(err, "could not parse build-arg %s", "DALEC_SKIP_TESTS")
```
Comment (MorrisLaw) [inline frontend/test_runner.go]: nit: no need to use a format specifier since it's just a hardcoded string; we could combine it into 1 string

Comment (cpuguy83) [reply frontend/test_runner.go]: Changed this up a bit.

File: frontend/test_runner.go (hunk 1 above)
Comment (cpuguy83) [reply frontend/test_runner.go]: Changed this up a bit.

File: targets/linux/rpm/distro/pkg.go [hunk 3]
```diff
@@ -68,29 +67,14 @@ func (c *Config) BuildPkg(ctx context.Context, client gwclient.Client, sOpt dale
 	return frontend.MaybeSign(ctx, client, st, spec, targetKey, sOpt, opts...)
 }
 
-// runTests runs the package tests
+// RunTests runs the package tests
 // The returned reference is the solved container state
-func (cfg *Config) RunTests(ctx context.Context, client gwclient.Client, spec *dalec.Spec, sOpt dalec.SourceOpts, ctr llb.State, targetKey string, opts ...llb.ConstraintsOpt) (gwclient.Reference, error) {
-	def, err := ctr.Marshal(ctx, opts...)
```
Comment (invidian) [inline targets/linux/rpm/distro/pkg.go:74]: Nice seeing that duplicated bits refactored!

File: test/linux_target_test.go [hunk 4]
```diff
@@ -3089,44 +3093,69 @@ func testLinuxPackageTestsFail(ctx context.Context, t *testing.T, cfg testLinuxC
...
-				assert.ErrorContains(t, err, v)
+		doTest := func(target string) func(t *testing.T) {
+			return func(t *testing.T) {
+				t.Parallel()
 
-				v = (&dalec.CheckOutputError{Path: "/", Kind: dalec.CheckFilePermissionsKind, Expected: "-rw-r--r--", Actual: "-rwxr-xr-x"}).Error()
-				assert.ErrorContains(t, err, v)
+				dir := t.TempDir()
+				f, err := os.OpenFile(filepath.Join(dir, "out.txt"), os.O_CREATE|os.O_RDWR, 0o600)
+				assert.NilError(t, err)
```
Comment (invidian) [inline test/linux_target_test.go:3115]: I'm not sure that's the good place for discussing this, but generally speaking, `assert` does not interrupt the execution of the function which often leads to panics and false-negatives (e.g. when f is nil), so maybe it would be better to use `require` instead, which gives clearer failure handling?

File: test/linux_target_test.go [hunk 5]
```diff
@@ -3089,44 +3093,69 @@ func testLinuxPackageTestsFail(ctx context.Context, t *testing.T, cfg testLinuxC
...
+					}
+				}
 
-				v = (&dalec.CheckOutputError{Path: "/some_symlink2", Kind: dalec.CheckFileLinkTargetPathKind, Expected: "/not-a-real-path2", Actual: "/usr/bin/a-thing-for-symlinking"}).Error()
-				assert.ErrorContains(t, err, v)
+				ctx := startTestSpan(ctx, t)
+				testEnv.RunTest(ctx, t, func(ctx context.Context, client gwclient.Client) {
+					sr := newSolveRequest(withSpec(ctx, t, spec), withBuildTarget(target))
+					_, err := client.Solve(ctx, sr)
+					assert.Assert(t, err != nil)
```
Comment (invidian) [inline test/linux_target_test.go:3133]: `require.NilError`?

File: test/linux_target_test.go [hunk 6]
```diff
@@ -3089,44 +3093,69 @@ func testLinuxPackageTestsFail(ctx context.Context, t *testing.T, cfg testLinuxC
...
+			(&dalec.CheckOutputError{Path: "/some_symlink3", Kind: dalec.CheckFileNotExistsKind, Expected: "exists=true", Actual: "exists=false"}).Error(),
+			(&dalec.CheckOutputError{Path: "/some_symlink4", Kind: dalec.CheckFileLinkTargetPathKind, Expected: "/incorrect-target", Actual: "/not-a-real-path4"}).Error(),
+			"step did not complete successfully: exit code: 42",
+			"stdout not hello",
+			"stderr not hello",
+		}
 
-				v := (&dalec.CheckOutputError{Path: "/non-existing-file", Kind: dalec.CheckFileNotExistsKind, Expected: "exists=true", Actual: "exists=false"}).Error()
-				assert.ErrorContains(t, err, v)
+		doTest := func(target string) func(t *testing.T) {
```
Comment (invidian) [inline test/linux_target_test.go]: Maybe?
```suggestion
		testsForTarget := func(target string) func(t *testing.T) {
```

File: frontend/test_runner.go [hunk 7]
```diff
@@ -23,180 +25,222 @@ import (
...
-		if err != nil {
-			return errors.Wrapf(err, "could not parse build-arg %s", "DALEC_SKIP_TESTS")
-		}
-		if skip {
-			Warn(ctx, client, llb.Scratch(), "Tests skipped due to build-arg DALEC_SKIP_TESTS="+skipVar)
-			return nil
+// RunTests runs the tests defined in the spec against the given the input state.
+// The result of this is either the provided `finalState` or a state that errors when executed with the errors produced by the tests.
+// The input state should be a runnable container with all dependencies already installed.
+func RunTests(ctx context.Context, client gwclient.Client, spec *dalec.Spec, finalState llb.State, target string, platform *ocispecs.Platform) llb.StateOption {
```
Comment (invidian) [inline frontend/test_runner.go:37]: If implementing `llb.StateOption`, do we actually need to take `finalState llb.State` as argument? Wouldn't `in llb.State` always be the final state?

File: cmd/frontend/main.go [hunk 8]
```diff
@@ -55,6 +55,9 @@ func main() {
 	case testrunner.CheckFilesCmdName:
 		args := flag.Args()[2:]
 		testrunner.CheckFilesCmd(args)
+	case frontend.TestErrorCmdName:
```
Comment (invidian) [inline cmd/frontend/main.go:58]: Why do we need a separate frontend command to handle printing errors? Couldn't this just run after we are finished running tests?

File: test/linux_target_test.go (hunk 4 above)
Comment (cpuguy83) [reply test/linux_target_test.go:3115]: `assert` here is not testify, its `gotest.tools/v3/asssert`.
`assert.NilError` does halt the test

File: test/linux_target_test.go (hunk 5 above)
Comment (cpuguy83) [reply test/linux_target_test.go:3133]: This is asserting that we have an error.

File: frontend/test_runner.go (hunk 7 above)
Comment (cpuguy83) [reply frontend/test_runner.go:37]: `finalState` here is because the input state is not always going to be the final state, e.g. for packaging targets such as jammy/deb and sysext targets.

File: cmd/frontend/main.go (hunk 8 above)
Comment (cpuguy83) [reply cmd/frontend/main.go:58]: This isn't just printing the errors, its exiting non-zero.
The reason we have this is so we can return an LLB state to buildkit that will error out and fail the build rather than returning an error to buildkit which messes up the dap debugger.

I actually have an alternative to this where each test assertion is actually run as an exec (`state.Run(...)`) but that seems to be triggering a bug in buildkit.
So I've settled for this case where we keep running the tests like we used to which requires un-lazying the build but still give buildkit a valid LLB state that produces an error.

File: test/linux_target_test.go (hunk 4 above)
Comment (invidian) [reply test/linux_target_test.go:3115]: Ah, that's confusing, TIL. Maybe we could remove testify assert completely to avoid this confusion. I'll have a look into that.

Comment (cpuguy83) [reply test/linux_target_test.go:3115]: Indeed, I opened an issue yesterday to track this, except I suggested switching to testify since that is a more well known library (of course either way seems fine to me).

File: frontend/test_runner.go (hunk 7 above)
Comment (invidian) [reply frontend/test_runner.go:37]: I think I still don't understand. Could you elaborate to help me graps how this works? So the input state could be worker with test dependencies, while final state would be where deb/sysext are? Do I get this right?

File: frontend/test_runner.go [hunk 9]
```diff
@@ -23,180 +25,222 @@ import (
...
 				}
-			}
 
-			return stderrors.Join(errs...)
-		})
+				return stderrors.Join(errs...)
+			})
+		}
+
+		err = group.Wait()
```
Comment (invidian) [inline frontend/test_runner.go]: I don't think it's the task for this PR, but I feel code in RunTests would benefit from the extensive test suite for all different behaviours, since there is really a lot of code and as a newcomer it's hard to figure out what it all does.

Or maybe alternatively some refactoring to better organize levels of abstraction in this function could help make it easier to understand and drive down the complexity would be sufficient.

File: frontend/test_runner.go (hunk 7 above)
Comment (cpuguy83) [reply frontend/test_runner.go:37]: The input state must always be a state that is ready to have test assertions executed.
deb/rpm/sysext are not that but we still want to run tests when building those targets (actually even the container state may have extra stuff due to test dependencies being installed).

In those cases we make the input state the container (ie package installed) + test dependencies.
The final state, which the test runner outputs, needs to be what we plan to output OR we need to handle that at the call site, which could also be fair (maybe a separate state option called Reset or something).

File: frontend/test_runner.go (hunk 9 above)
Comment (cpuguy83) [reply frontend/test_runner.go]: I think we have a large amount of coverage for the test runner within the integration test suite as it is.
Its possible an audit may come up with some case that's not tested but we do cover both positive and negative case in there to validate the test runner is correct and we also rely on the test runner for some of the rest of the test suite.

File: frontend/test_runner.go (hunk 7 above)
Comment (cpuguy83) [reply frontend/test_runner.go:37]: That said, I agree the code itself is complex.
I factored out a bunch of this in #898 but ran into some issues that need to be dug into further and may require fixes in buildkit (which would also delay actually being able to refactor since the test runner wouldn't be usable on older buildkits).

Comment (cpuguy83) [reply frontend/test_runner.go:37]: I factored out a few things to split this loop up in to smaller chunks, hope this helps (all in the 2nd commit).

Comment (invidian) [reply frontend/test_runner.go:37]: Thanks for the explanation. It makes sense now :)
